File: packages/domain-types/src/events/orchestration.ts
/**
 * Pipeline Orchestration domain events.
 *
 * @see ddd-target.md §4.7
 */

import type { TenantId } from "../tenant.js";
import { type DomainEvent, createDomainEvent } from "./base.js";

// -- StageStarted -----------------------------------------------------------

export interface StageStartedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly attemptNumber: number;
  readonly startedAt: string;
}

export type StageStarted = DomainEvent<"StageStarted", StageStartedPayload>;

export function createStageStarted(
  tenantId: TenantId,
  payload: StageStartedPayload,
): StageStarted {
  return createDomainEvent("StageStarted", tenantId, payload);
}

// -- StageCompleted ---------------------------------------------------------

export interface StageCompletedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly finishedAt: string;
  readonly durationMs: number;
}

export type StageCompleted = DomainEvent<"StageCompleted", StageCompletedPayload>;

export function createStageCompleted(
  tenantId: TenantId,
  payload: StageCompletedPayload,
): StageCompleted {
  return createDomainEvent("StageCompleted", tenantId, payload);
}

// -- StageFailed ------------------------------------------------------------

export interface StageFailedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly errorCode: string;
  readonly errorMessage: string;
  readonly retryable: boolean;
  readonly attemptNumber: number;
}

export type StageFailed = DomainEvent<"StageFailed", StageFailedPayload>;

export function createStageFailed(
  tenantId: TenantId,
  payload: StageFailedPayload,
): StageFailed {
  return createDomainEvent("StageFailed", tenantId, payload);
}

// -- StageExhausted ---------------------------------------------------------

export interface StageExhaustedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly attemptCount: number;
  readonly maxAttempts: number;
}

export type StageExhausted = DomainEvent<"StageExhausted", StageExhaustedPayload>;

export function createStageExhausted(
  tenantId: TenantId,
  payload: StageExhaustedPayload,
): StageExhausted {
  return createDomainEvent("StageExhausted", tenantId, payload);
}

// -- StageReset -------------------------------------------------------------

export interface StageResetPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly resetAttempts: boolean;
  readonly resetAt: string;
}

export type StageReset = DomainEvent<"StageReset", StageResetPayload>;

export function createStageReset(
  tenantId: TenantId,
  payload: StageResetPayload,
): StageReset {
  return createDomainEvent("StageReset", tenantId, payload);
}

// -- StageBlocked -----------------------------------------------------------

export interface StageBlockedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly blockedBy: readonly string[];
}

export type StageBlocked = DomainEvent<"StageBlocked", StageBlockedPayload>;

export function createStageBlocked(
  tenantId: TenantId,
  payload: StageBlockedPayload,
): StageBlocked {
  return createDomainEvent("StageBlocked", tenantId, payload);
}

// -- StageSkipped -----------------------------------------------------------

export interface StageSkippedPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly reason: string;
}

export type StageSkipped = DomainEvent<"StageSkipped", StageSkippedPayload>;

export function createStageSkipped(
  tenantId: TenantId,
  payload: StageSkippedPayload,
): StageSkipped {
  return createDomainEvent("StageSkipped", tenantId, payload);
}

// -- StageCanceled ----------------------------------------------------------

export interface StageCanceledPayload {
  readonly jobId: string;
  readonly stage: string;
  readonly canceledAt: string;
  readonly reason?: string;
}

export type StageCanceled = DomainEvent<"StageCanceled", StageCanceledPayload>;

export function createStageCanceled(
  tenantId: TenantId,
  payload: StageCanceledPayload,
): StageCanceled {
  return createDomainEvent("StageCanceled", tenantId, payload);
}

File: workers/automation/src/jobctrl/domain/events/orchestration.py
"""Pipeline Orchestration domain events.

See ddd-target.md §4.7.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass

from jobctrl.domain.tenant import TenantId
from jobctrl.domain.events.base import DomainEvent, create_domain_event


@dataclass(frozen=True)
class StageStartedPayload:
    job_id: str
    stage: str
    attempt_number: int
    started_at: str


def create_stage_started(tenant_id: TenantId, payload: StageStartedPayload) -> DomainEvent:
    return create_domain_event("StageStarted", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageCompletedPayload:
    job_id: str
    stage: str
    finished_at: str
    duration_ms: int


def create_stage_completed(tenant_id: TenantId, payload: StageCompletedPayload) -> DomainEvent:
    return create_domain_event("StageCompleted", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageFailedPayload:
    job_id: str
    stage: str
    error_code: str
    error_message: str
    retryable: bool
    attempt_number: int


def create_stage_failed(tenant_id: TenantId, payload: StageFailedPayload) -> DomainEvent:
    return create_domain_event("StageFailed", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageExhaustedPayload:
    job_id: str
    stage: str
    attempt_count: int
    max_attempts: int


def create_stage_exhausted(tenant_id: TenantId, payload: StageExhaustedPayload) -> DomainEvent:
    return create_domain_event("StageExhausted", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageResetPayload:
    job_id: str
    stage: str
    reset_attempts: bool
    reset_at: str


def create_stage_reset(tenant_id: TenantId, payload: StageResetPayload) -> DomainEvent:
    return create_domain_event("StageReset", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageBlockedPayload:
    job_id: str
    stage: str
    blocked_by: tuple[str, ...] = ()


def create_stage_blocked(tenant_id: TenantId, payload: StageBlockedPayload) -> DomainEvent:
    return create_domain_event("StageBlocked", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageSkippedPayload:
    job_id: str
    stage: str
    reason: str


def create_stage_skipped(tenant_id: TenantId, payload: StageSkippedPayload) -> DomainEvent:
    return create_domain_event("StageSkipped", tenant_id, asdict(payload))


@dataclass(frozen=True)
class StageCanceledPayload:
    job_id: str
    stage: str
    canceled_at: str
    reason: str | None = None


def create_stage_canceled(tenant_id: TenantId, payload: StageCanceledPayload) -> DomainEvent:
    return create_domain_event("StageCanceled", tenant_id, asdict(payload))
